src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs

```diff
@@ -459,19 +516,24 @@ internal static void EnsureGroupPreservationColumn(IHostEnvironment env, ref IDa
                     // of column types. It used to be HashJoin, but we should probably extend Hash
                     // instead of having two hash transformations.
                     var origStratCol = samplingKeyColumn;
-                    int tmp;
-                    int inc = 0;
-
-                    // Generate a new column with the hashed samplingKeyColumn.
-                    while (data.Schema.TryGetColumnIndex(samplingKeyColumn, out tmp))
-                        samplingKeyColumn = string.Format("{0}_{1:000}", origStratCol, ++inc);
+                    samplingKeyColumn = data.Schema.GetTempColumnName(samplingKeyColumn);
                     HashingEstimator.ColumnOptions columnOptions;
                     if (seed.HasValue)
                         columnOptions = new HashingEstimator.ColumnOptions(samplingKeyColumn, origStratCol, 30, (uint)seed.Value);
                     else
                         columnOptions = new HashingEstimator.ColumnOptions(samplingKeyColumn, origStratCol, 30);
                     data = new HashingEstimator(env, columnOptions).Fit(data).Transform(data);
                 }
+                else
+                {
+                    if (!data.Schema[samplingKeyColumn].IsNormalized() && (type == NumberDataViewType.Single || type == NumberDataViewType.Double))
```

Comment:
@Ivanidzo4ka -- and on top of all of this, you spontaneously fixed a potential hole. amazing